### ISSUE TITLE
Rewrite in Typescript

### ISSUE THREAD

**Issue by axic** [ethereum/solc-js#287]

**Comment by axic** [comment]
@s1na would you be interested in this?

**Comment by s1na** [comment]
Yeah I would. How do you want the transition to happen? If the whole source (+ tests) are rewritten in TS, I imagine it'll be a somewhat big PR.

**Comment by axic** [comment]
Not sure yet, perhaps we could start with changing the tests first, then some of the helper libs and finally the rest?

**Comment by MambaDev** [comment]
ping 🏓 https://github.com/ethereum/solc-js/pull/442

**Comment by sam-goodwin** [comment]
Would love to see this implemented. Was it de-prioritized?

**Comment by cameel** [comment]
See https://github.com/ethereum/solc-js/pull/442#issuecomment-912766052. No one from the team is currently working on this and almost all the work is going on in the main C++ repo.

We have a few open PRs from the community (#205, #298, #442) but they are all pretty old and probably already out of date. The ones from @MicahZoltu look like a good first step so if someone is willing to take over, finish remaining loose ends and bring them up to date, I can help by reviewing them. Probably a lot could also be salvaged from #298 as long as it would be a series of smaller PRs and not one huge PR changing everything.

Given how old these are, I think it would also be fine for someone to just ignore them and start a fresh PR. That might be more motivating than trying to revive code written by someone else.

**Comment by cameel** [comment]
Sounds good to me. @MambaDev, if you want to try that, feel free to take over the PRs already submitted by @MicahZoltu.

BTW, if you do any mass renaming, please make sure the renames are in a separate commit without any other changes. This makes it **much** easier to see in the review what was just moved around and what was actually changed. It also generally helps with conflicts since git can properly see whole files as moved.

**Comment by StephenSLI** [comment]
That is all fine @cameel, no worries. 

I will start staging it, e.g setup the infrastructure (building), renaming etc in one pull request and then allow building on that.

**Comment by StephenSLI** [comment]
I have gone and started up the first pull request to start the base work for moving into the Typescript space, its #566 and its all open to conversation.

**Comment by StephenSLI** [comment]
#567 Pull request has just been open to start the process of moving to Eslint.

**Comment by cameel** [comment]
Just a general update: #566 by @stephensli has been merged recently so we're finally on TS! It introduced some backwards-incompatibility but this is fixed in #595 and #596 so it will not actually show up in a release.

The next step is adding types for development and then types for the exported functions.

**Comment by GrandSchtroumpf** [comment]
Hello, maybe it can help. Back in the days, I created the types for the input/output compilation for remix-plugin.
You can find the types here : https://github.com/ethereum/remix-plugin/tree/master/packages/api/src/lib/compiler/type 
It might not be up-to-date though (version 0.6.x), but it's a good base I think.

**Comment by stephensli** [comment]
HI @GrandSchtroumpf,

exposed types have been added here: https://github.com/ethereum/solc-js/blob/540a9643cd7974e329287ebf8d4c2c70d744c11d/common/types.ts pending a existing PR has to go in first. These are currently in review.

**Comment by GrandSchtroumpf** [comment]
Hi @stephensli my link above describe the types for the input (language, source, settings) and output (abi, ...) of the `compile` method based on the solidity [input/output documentation](https://docs.soliditylang.org/en/v0.8.14/using-the-compiler.html#input-description).
I can create a separated PR for that if it's fine for you

**Comment by roninjin10** [comment]
I put up a pr to release types on npm here https://github.com/ethereum/solc-js/pull/693

**Comment by EdouardBougon** [comment]
Any update on types release ?
